src/freemindlatex/compilation_server.py

```diff
@@ -0,0 +1,292 @@
+      generated .pdf file.
+
+  Raises:
+    BibtexCompilationError: when bibtex compilation encounters some errors
+      or warnings
+  """
+  proc = subprocess.Popen(
+    ["bibtex",
+     filename_prefix], cwd=working_dir, stdout=subprocess.PIPE,
+    stderr=subprocess.PIPE)
+  stdout, _ = proc.communicate()
+  if proc.returncode != 0:
+    raise BibtexCompilationError(stdout)
+
+
+def _ParseNodeIdAndErrorMessageMapping(
+    latex_content, latex_compilation_error_msg):
+  """Parse the latex compilation error message, to see which frames have errors.
+
+  Args:
+    latex_content: the mindmap.tex file content, including frames'
+    node markers. We use it to extract mappings between line numbers and frames
+    latex_compilation_error_msg: the latex compilation error message, containing
+      line numbers and error messages.
+
+  Returns:
+    A map of frame IDs and the compilation errors within it. For example:
+    { "node12345" : ["nested too deep"] }
+  """
+  result = collections.defaultdict(list)
+
+  lineno_frameid_map = {}
+
+  frame_node_id = None
+  for line_no, line in enumerate(latex_content.split("\n")):
+    line_no = line_no + 1
+    if line.startswith("%%frame: "):
+      frame_node_id = re.match(r'%%frame: (.*)%%', line).group(1)
+
+    if frame_node_id is not None:
+      lineno_frameid_map[line_no] = frame_node_id
+
+  lineno_and_errors = []
+  error_message = None
+  for line in latex_compilation_error_msg.split("\n"):
+    if line.startswith("! "):
+      error_message = line[2:]
+    mo = re.match(r'l.(\d+)', line)
+    if mo is not None:
+      lineno_and_errors.append((int(mo.group(1)), error_message))
+
+  for lineno, error in lineno_and_errors:
+    frame_id = lineno_frameid_map[lineno]
+    result[frame_id].append(error)
+
+  return result
+
+
+def _LatexCompileOrTryEmbedErrorMessage(org, work_dir):
+  """Try compiling. If fails, try embedding error messages into the frame.
+
+  Args:
+    org: the in-memory slides organization
+    work_dir: Directory containing the running files: mindmap.mm,
+      and the image files.
+
+  Returns:
+    A compilation_service.LatexCompilationResponse object.
+  """
+  output_tex_file_loc = os.path.join(work_dir, "mindmap.tex")
+  org.OutputToBeamerLatex(output_tex_file_loc)
+
+  # First attempt
+  result = _CompileLatexAtDir(work_dir, "slides.tex")
+
+  if result.status == "SUCCESS":
+    return result
+
+  # Second attempt
+  frame_and_error_message_map = _ParseNodeIdAndErrorMessageMapping(
+    result.source_code, result.compilation_log)
+  org.LabelErrorsOnFrames(frame_and_error_message_map)
+  org.OutputToBeamerLatex(output_tex_file_loc)
+
+  second_attempt_result = _CompileLatexAtDir(work_dir, "slides.tex")
+  if second_attempt_result.status == "SUCCESS":
+    result.status = "EMBEDDED"
+
+  else:
+    result.status = "CANNOTFIX"
+  return result
+
+
+class BibtexCompilationError(Exception):
+  pass
+
+
+def _CompileInWorkingDirectory(work_dir):
```

Comment:
I should merge this function into the function below. Because this is the main logic for compilation. The that function becomes too long, extract the directory preparation part.